### ISSUE TITLE
tl.gather result zeroed when assigned to named intermediate in @triton.jit kernel (Arc B580, Triton 3.7.0)

### ISSUE THREAD

**Issue by hammurabi-coder** [intel/intel-xpu-backend-for-triton#6511]
## Bug Description

Assigning `tl.gather` to a named intermediate variable produces all zeros when the result is subsequently used in `tl.where` on the XPU backend (Intel Arc B580, Triton 3.7.0).

## Steps to Reproduce

```python
import torch, triton, triton.language as tl
DEVICE = torch.device('xpu')

@triton.jit
def fwht4_broken(x_ptr, D: tl.constexpr, BS: tl.constexpr):
    row = tl.program_id(0)
    off = tl.arange(0, BS)
    x = tl.load(x_ptr + row * D + off)
    g = tl.gather(x, off ^ 1, axis=0)   # stored to named var → zeros on XPU
    x = tl.where((off & 1) != 0, x - g, x + g)
    tl.store(x_ptr + row * D + off, x)

# Test: input [1,2,3,4] -> expected FWHT output [10,-2,-4,0]
# Broken kernel produces [10, 2, 4, 0] — upper elements wrong sign
# "Fixed" by using gather directly inline (no named intermediate)
```

## Expected vs Actual

- **Expected**: `tl.gather` returns correct partner values from the vector
- **Actual**: Stored `tl.gather` result is all zeros when read in subsequent `tl.where`
- **Inline usage works**: `tl.where(mask, tl.gather(x, off^1) - x, x + tl.gather(x, off^1))` produces correct output

## Environment

- Intel Arc B580 (Battlemage, 12GB VRAM)
- WSL2 Ubuntu 22.04
- PyTorch 2.5.1+ (xpu support)
- Triton 3.7.0
- intel-xpu-backend-for-triton (pytorch.org whl)

## Additional Notes

- Does NOT affect CUDA backend — same code works correctly on CUDA
- Appears to be a register allocation/materialization issue in the XPU JIT compiler
- The gather appears to execute correctly (no error), but the stored result reads as zeros when used later
- Workaround: use `tl.gather` directly inline in `tl.where` (appears twice, compiler CSEs it)

**Comment by mieshkiwrk** [comment]
Hi @hammurabi-coder,
I'm confused due to few things

1. I don't see any zeros results when using `tl.gather` with subsequent `tl.where`, named intermediate variable vs inline usage shouldn't matter because it produces identical IR.

2. This provided kernel only calculates first stage (step=1) while for input size 4 it should be `log2(4) = 2`, 
so for `[1, 2, 3, 4]` expected output should be `[3, -1, 7, -1]`, then in step=2 -> `[10, -2, -4, 0]`

3. Provided `tl.where` operation mentioning one works while other not have different operation order
```python
# Assumed invalid
g = tl.gather(x, off ^ 1, axis=0)   # stored to named var → zeros on XPU
x = tl.where((off & 1) != 0, x - g, x + g)

# Assumed working
tl.where(mask, tl.gather(x, off^1) - x, x + tl.gather(x, off^1))

# Comparing them:
x = tl.where((off & 1) != 0, x - g, x + g)
x = tl.where((off & 1) != 0, g - x, x + g)

```
---------------------------

So summarizing I don't see any issue there, can you check below reproducer and if you still believe there's some issue at the backend side please provide self-contained reproducer that highlight the problem


Here's full reproducer I've used (verified both PVC and BMG):
```python
import math
import torch
import triton
import triton.language as tl


@triton.jit
def fwht4(x_ptr, BLOCK_SIZE: tl.constexpr, STEP: tl.constexpr):
    off = tl.arange(0, BLOCK_SIZE)
    x = tl.load(x_ptr + off)
    g = tl.gather(x, off ^ STEP, axis=0)

    # INVALID
    # x = tl.where((off & STEP) != 0, x - g, x + g)

    # CORRECT
    x = tl.where((off & STEP) != 0, g - x, x + g)

    tl.store(x_ptr + off, x)


x = torch.tensor([1.0, 2.0, 3.0, 4.0], device='xpu')
input_size = len(x)

steps = math.log2(input_size)
assert steps.is_integer()

for i in range(0, int(steps)):
    step = 2 ** i
    fwht4[(1,)](x, BLOCK_SIZE=len(x), STEP=step)
    print(f"[Step={step}] Result:   {x.tolist()}")

print("Expected: [10.0, -2.0, -4.0, 0.0]")
```

Environment:
```
# scripts/capture-hw-details.sh

LIBIGC1_VERSION=2.27.10-1222
LEVEL_ZERO_VERSION=1.26.2-1~24.04
AGAMA_VERSION=1222
TORCH_VERSION=2.12.0a0+git29094e4
COMPILER_VERSION=2025.3.2
```

**Comment by hammurabi-coder** [comment]
Thank you for the thorough analysis. You're correct on all points — not a real compiler bug. The actual  problem was the subtraction operand order (x - g vs  g - x), which I conflated with the gather behavior  during debugging. The minimal repro was also  incomplete (one stage instead of log2(n)). Closing this as user error. Apologies for the noise.
